PP-489: Logs should have host and interface information

sid2364 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-489](https://pbspro.atlassian.net/browse/PP-489)**

#### Problem description
* Adding host and interface information to logs.

#### Solution description
* Collecting relevant information from system about host and interfaces to add to beginning of log, i.e. when log is opened.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

agrawalravi90 commented:
I think you used the small 's' while running your git commit, so you ended up "signing off" your commit instead of "signing" your commit with your GPG key. Just use "git commit --amend -S" (capital 'S') and you should be good.
Also, that's an incredibly long title, more than half of it overflowed into the commit body, might confuse people. I'd recommend just truncating it, people can find out the title from the link to the ticket. Or just mention the whole title separately in the commit body.

sid2364 commented:
@prakashcv13 Yes, [here](https://pbspro.atlassian.net/wiki/display/PD/PP-489%3A+MOM+should+log+her+hostname) is the link.
@agrawalravi90 Just went back and checked. Had used the small 's' by mistake. Corrected.

hirenvadalia commented:
@sid2364 Any reason why you are not writing PTL automated tests for this PR?

sid2364 commented:
@hirenvadalia Here, we're making system calls to get host name and interface information. It can be verified directly from the output in the log files.

prakashcv13 commented:
@sid2364, we can write PTL tests to check for these logs.

hirenvadalia commented:
@sid2364 As I can see results of manual testing for this PR you are checking mom logs for verification which we can easily automate using PTL (as @prakashcv13 told). In case, if you need help please feel free to ask, "we are happy to help you 😄"

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -334,6 +374,265 @@ log_init(void)
+	strcpy(temp, "pbs_leaf_name=");
+	if(pbs_conf.pbs_leaf_name){ 
+		strcat(temp, pbs_conf.pbs_leaf_name);
+		strcat(temp, ";");
```
Use strncat() to be safe.

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -334,6 +374,265 @@ log_init(void)
+	/* To add mom node name, if set */
+	strcpy(temp, "pbs_mom_node_name=");
+	if(pbs_conf.pbs_mom_node_name)
+		strcat(temp, pbs_conf.pbs_mom_node_name);
```
Use strncat() to be safe.

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -334,6 +374,265 @@ log_init(void)
+
+	ret = getifaddrs(&ifp);
+	if ((ret != 0) || (ifp == NULL)) {
+		fprintf(stderr, "ERROR: Failed to obtain interface names.\n");
```
I think you want to log this message, not print it.

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -334,6 +374,265 @@ log_init(void)
+		snprintf(tbuf, LOG_BUF_SIZE, "%s interface ", inet_family);
+
+		if (!hostnames)
+			continue;
```
Move this check immediately after hostnames is assigned. No need to call get_sa_family() and snprintf() since tbuf() isn't going to get logged.

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -334,6 +374,265 @@ log_init(void)
+		strcat(tbuf, ": ");
+		for (i = 0; hostnames[i]; i++){
+			strcat(tbuf, hostnames[i]);
+			strcat(tbuf, " ");
```
Use strncat() where possible.

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -334,6 +374,265 @@ log_init(void)
+
+	addrlistp = (IP_ADAPTER_ADDRESSES *)malloc(size);
+	if (!addrlistp) {
+		fprintf(stderr, "ERROR: Out of memory.\n");
```
Replace all instances of fprintf() with calls to log_record(). There are others I am not commenting on.

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -334,6 +374,265 @@ log_init(void)
+			return 1;
+		}
+	}
+	ret = GetAdaptersAddresses(AF_UNSPEC, GAA_FLAG_INCLUDE_PREFIX, NULL, addrlistp, &size);
```
This line should be within the above if statement. No need to call it again if we did not encounter an ERROR_BUFFER_OVERFLOW.

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -334,6 +374,265 @@ log_init(void)
+			strcat(tbuf, ": ");
+			for (i = 0; hostnames[i]; i++){
+				strcat(tbuf, hostnames[i]);
+				strcat(tbuf, " ");
```
Use strncat().

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -334,6 +374,265 @@ log_init(void)
+			strcat(tbuf, ": ");
+			for (i = 0; hostnames[i]; i++){
+				strcat(tbuf, hostnames[i]);
+				strcat(tbuf, " ");
```
Use strncat().

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -464,6 +763,13 @@ log_open_main(char *filename, char *directory, int silent)
+
+			log_add_debug_info();
+
+			if((ret = log_add_if_info()) == 1){
```
Why declare and assign ret if you never reference it once it is set?

mike0042 commented on src/lib/Libutil/hnls.c:
```diff
@@ -0,0 +1,170 @@
+#endif
+
+void
+free_if_hostnames(char **names)
```
Is there a reason you're not delcaring the functions in this file in libutil.h?

mike0042 commented:
I found an EDD here: https://pbspro.atlassian.net/wiki/display/PD/PP-489%3A+MOM+should+log+her+hostname

I did not find an announcement on the community forum requesting reviews. Please post one.

You should add interface definitions for the library calls you are adding as well. Please note my comments at the very top of my review regarding where these functions live and what they return.

You're pull request should include tests. You checked the box indicating you added manual tests, but I did not find any in the ticket. Why can't you add a test that looks for the strings being logged?

sid2364 commented:
@mike0042 Will address comments and update the code accordingly. In the JIRA ticket, I've added the log outputs that include the newly added information (hostname and interface info), which I thought would account for manual tests/checks.

@hirenvadalia Definitely will need some help to start off with PTL!

sid2364 commented:
@mike0042 I think you're right; this code should be in Libnet as opposed to Libutil. Will make this change too.

sid2364 commented:
@mike0042 There was a post on the community forum by @nithinj about this. It can be found ([here]http://community.pbspro.org/t/pp-489-mom-should-log-her-hostname/360).

Also, I've addressed the comments that were put for the last commit.

mike0042 commented on src/lib/Libnet/hnls.c:
```diff
@@ -0,0 +1,442 @@
+}
+
+void
+get_if_info(struct log_net_info *ni, char *msg){
```
This function should be declared in a header file. Please place the open brace on its own line. Please add doxygen headers to the functions in this file.

sid2364 replied:
Have made the changes you asked for.

sid2364 commented:
EDD has been updated. You can find it [here](https://pbspro.atlassian.net/wiki/display/PD/PP-489%3A+Daemons+should+log+hostname+and+interface+information).

sid2364 commented:
@mike0042 Have addressed your comments and updated the code.

sid2364 commented:
Have rebased and squashed the commits.
